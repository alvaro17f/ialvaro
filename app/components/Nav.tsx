import Link from "next/link";

const links = [
	{
		label: "Home",
		route: "/",
	},
	{
		label: "About",
		route: "/about",
	},
];

export default function Nav() {
	return (
		<header>
			<nav>
				<ul>
					{links.map(({ label, route }) => (
						<li key={route}>
							<Link href={route}>{label}</Link>
						</li>
					))}
				</ul>
			</nav>
		</header>
	);
}
