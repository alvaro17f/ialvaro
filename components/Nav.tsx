"use client";
import { m, LazyMotion, domAnimation, AnimatePresence } from "framer-motion";
import { useEffect, useRef, useState } from "react";

const links = [
	{
		label: "Home",
		route: "home",
	},
	{
		label: "Biography",
		route: "biography",
	},
	{
		label: "Skills",
		route: "skills",
	},
	{
		label: "Experience",
		route: "experience",
	},
	// {
	// 	label: "Portfolio",
	// 	route: "portfolio",
	// },
	{
		label: "Resume",
		route: "resume",
	},
	{
		label: "Contact",
		route: "contact",
	},
];

export default function Nav() {
	const [scrollPosition, setScrollPosition] = useState(0);
	const [isOpen, setIsOpen] = useState(false);
	const ref = useRef(null);

	useEffect(() => {
		const updatePosition = () => {
			setScrollPosition(window.scrollY);
		};

		window.addEventListener("scroll", updatePosition);

		return () => window.removeEventListener("scroll", updatePosition);
	}, []);

	const scroller = (route: string) =>
		document.getElementById(route)?.scrollIntoView({ behavior: "smooth" });

	return (
		<>
			<LazyMotion features={domAnimation}>
				<AnimatePresence>
					{scrollPosition > 700 && (
						<m.nav
							className="fixed z-50 w-full bg-alvaro-base rounded-xl "
							initial={{ opacity: 0 }}
							animate={{ opacity: 1 }}
							exit={{ opacity: 0, transition: { duration: 0.3 } }}
							transition={{ duration: 1.5 }}
						>
							<div className="px-4 mx-auto max-w-7xl sm:px-6 lg:px-8">
								<div className="flex items-center justify-between h-16">
									<div className="flex items-center">
										<div className="flex-shrink-0">
											<button
												onClick={() => {
													scroller("home");
													isOpen && setIsOpen(!isOpen);
												}}
											>
												<img
													src="/images/ialvaro.svg"
													alt="logo"
													className="h-10"
												/>
											</button>
										</div>
										<div className="hidden md:block">
											<div className="flex items-baseline ml-10 space-x-4">
												{links.map(({ label, route }) => (
													<m.div
														key={route}
														initial={{ scale: 1 }}
														whileHover={{ scale: 1.2 }}
														whileTap={{ scale: 1 }}
													>
														<button
															key={route}
															onClick={() => scroller(route)}
															className="px-3 py-2 text-sm font-medium rounded-md cursor-pointer text-alvaro-white hover:bg-alvaro-danger hover:text-alvaro-dark"
															aria-label={`${label}-desktop`}
														>
															{label}
														</button>
													</m.div>
												))}
											</div>
										</div>
									</div>
									<div className="flex -mr-2 md:hidden">
										<button
											onClick={() => setIsOpen(!isOpen)}
											type="button"
											className="inline-flex items-center justify-center p-2 rounded-md text-alvaro-white focus:outline-none focus:ring-2 focus:ring-offset-2"
											aria-controls="mobile-menu"
											aria-expanded="false"
										>
											<span className="sr-only">menu-mobile</span>
											{!isOpen ? (
												<svg
													className="block w-6 h-6"
													xmlns="http://www.w3.org/2000/svg"
													fill="none"
													viewBox="0 0 24 24"
													stroke="currentColor"
													aria-hidden="true"
												>
													<path
														strokeLinecap="round"
														strokeLinejoin="round"
														strokeWidth="2"
														d="M4 6h16M4 12h16M4 18h16"
													/>
												</svg>
											) : (
												<svg
													className="block w-6 h-6"
													xmlns="http://www.w3.org/2000/svg"
													fill="none"
													viewBox="0 0 24 24"
													stroke="currentColor"
													aria-hidden="true"
												>
													<path
														strokeLinecap="round"
														strokeLinejoin="round"
														strokeWidth="2"
														d="M6 18L18 6M6 6l12 12"
													/>
												</svg>
											)}
										</button>
									</div>
								</div>
							</div>

							{isOpen && (
								<div ref={ref} className="px-2 pt-2 pb-3 space-y-1 sm:px-3">
									<m.div
										className="md:hidden"
										initial={{ x: 10, opacity: 0 }}
										animate={{ x: 0, opacity: 1 }}
										exit={{ opacity: 0 }}
										transition={{ duration: 1 }}
									>
										{links.map(({ label, route }) => (
											<button
												key={route}
												onClick={() => {
													setIsOpen(!isOpen);
													scroller(route);
												}}
												className="block px-3 py-2 text-base font-medium text-white rounded-md hover:bg-alvaro-muted"
												aria-label={`${label}-mobile`}
											>
												{label}
											</button>
										))}
									</m.div>
								</div>
							)}
						</m.nav>
					)}
				</AnimatePresence>
			</LazyMotion>
		</>
	);
}
