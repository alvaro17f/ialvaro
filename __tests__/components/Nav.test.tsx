import Nav from "@/components/Nav";
import { afterEach, beforeEach, describe, expect, it, vi } from "vitest";
import {
	render,
	screen,
	within,
	fireEvent,
	cleanup,
	waitFor,
} from "@testing-library/react";

describe("<Nav />", () => {
	beforeEach(() => {
		const mockIntersectionObserver = vi.fn();
		mockIntersectionObserver.mockReturnValue({
			observe: () => null,
			unobserve: () => null,
			disconnect: () => null,
		});
		window.IntersectionObserver = mockIntersectionObserver;
		render(<Nav />);
	});
	afterEach(cleanup);

	it("should match the snapshot", () => {
		const { container } = render(<Nav />);
		expect(container).toMatchSnapshot();
	});

	it("should render the navbar logo", async () => {
		fireEvent.scroll(window, { target: { scrollY: 800 } });
		await waitFor(() => {
			const logo = screen.getByAltText("logo");
			expect(logo).toBeTruthy();
		});
	});

	it("should render links", async () => {
		fireEvent.scroll(window, { target: { scrollY: 800 } });
		await waitFor(() => {
			const navigation = within(screen.getByRole("navigation"));
			expect(navigation.getByLabelText(/home/i)).toBeDefined();
			expect(navigation.getByLabelText(/biography/i)).toBeDefined();
			expect(navigation.getByLabelText(/skills/i)).toBeDefined();
			expect(navigation.getByLabelText(/experience/i)).toBeDefined();
			expect(navigation.getByLabelText(/cv/i)).toBeDefined();
			expect(navigation.getByLabelText(/contact/i)).toBeDefined();
		});
	});

	it("should display the navbar menu when clicking on the mobile menu button", async () => {
		fireEvent.scroll(window, { target: { scrollY: 800 } });
		fireEvent.click(screen.getByLabelText(/menu-mobile/i));

		await waitFor(() => {
			const navigation = within(screen.getByRole("navigation"));
			expect(navigation.getByLabelText(/home-mobile/i)).toBeDefined();
			expect(navigation.getByLabelText(/biography-mobile/i)).toBeDefined();
			expect(navigation.getByLabelText(/skills-mobile/i)).toBeDefined();
			expect(navigation.getByLabelText(/experience-mobile/i)).toBeDefined();
			expect(navigation.getByLabelText(/cv-mobile/i)).toBeDefined();
			expect(navigation.getByLabelText(/contact-mobile/i)).toBeDefined();
		});
	});

	it("should close the navbar menu when a link is clicked", async () => {
		fireEvent.scroll(window, { target: { scrollY: 800 } });
		fireEvent.click(screen.getByLabelText(/menu-mobile/i));
		await waitFor(() => {
			const navigation = within(screen.getByRole("navigation"));
			fireEvent.click(navigation.getByLabelText(/home-mobile/i));
		});
		expect(screen.queryByLabelText(/home-mobile/i)).toBeNull();
	});

	it("should allow button to be clicked", () => {
		fireEvent.scroll(window, { target: { scrollY: 800 } });
		fireEvent.click(screen.getByAltText("logo"));
	});
});
