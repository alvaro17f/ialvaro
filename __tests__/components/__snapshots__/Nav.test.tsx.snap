// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Nav /> > should match the snapshot 1`] = `<div />`;
